Invoice CRUD

michaelr1300 opened the pull request:
### Migrations
- invoice
- invoice_book

### User Flow

1. User membuat faktur (invoice) lewat halaman add_invoice
2. Mengisi data faktur termasuk memilih customer dan buku
3. Faktur yang sudah dibuat ber status waiting dan masih bisa dilakukan perubahan (untuk jaga2 jika user ada kesalahan saat input data faktur)
4. Faktur yang sudah tidak perlu diubah lagi disetujui lewat halaman index_invoice
5. Faktur yang sudah disetujui tidak bisa diubah lagi
6. Buku yang diperlukan akan dikirim ke bagian gudang untuk disiapkan (next Pull Request)

### Technical Stuffs

1. Nomor Faktur di-generate otomatis di fungsi `get_last_invoice_number`.
Format nomor faktur
Digit pertama jenis faktur (**T**unai **K**redit **O**nline **S**howroom)
6 Digit tanggal pembuatan faktur (YYMMDD)
6 Digit urutan nomor faktur per jenis faktur

2. Data dropdown customer diambil dari database tabel customer. Data ditambah lewat halaman /customer atau bisa langsung lewat faktur
3. Dropdown buku di halaman tambah faktur hanya menampilkan buku yang stocknya lebih dari 0. (function `get_ready_book_list` di invoice_model)
4. Buku tidak bisa melebihi stock

michaelr1300 committed:
Invoice CRUD

- Migration
- Controller
- Model
- View
- Helper

michaelr1300 commented on application/database/migrations/20210501145022_Invoice_book.php:
```diff
@@ -0,0 +1,48 @@
+                    'constraint' => 10,
+                    'null' => TRUE
+                ],
+                'royalty' => [
```
Kolom untuk fitur ke depan, sementara belum dipakai

michaelr1300 commented on application/database/migrations/20210501151140_Invoice.php:
```diff
@@ -0,0 +1,96 @@
+                        'constraint' => 100,
+                        'null' => TRUE
+                    ],
+                    'total_weight' => [
```
Kolom untuk fitur ke depan, sementara belum dipakai

michaelr1300 commented on application/modules/invoice/controllers/invoice.php:
```diff
@@ -0,0 +1,293 @@
+        }
+    }
+
+    public function action($id, $invoice_status)
```
Function untuk menyetujui / menolak faktur (tombol aksi di invoice_index)

michaelr1300 commented on application/modules/invoice/models/invoice_model.php:
```diff
@@ -0,0 +1,326 @@
+{
+    public $per_page = 10;
+
+    public function validate_invoice()
```
Validation menggunakan ajax seperti /customer

michaelr1300 commented on application/modules/invoice/models/invoice_model.php:
```diff
@@ -0,0 +1,326 @@
+        return $stock;
+    }
+
+    public function get_ready_book_list()
```
Menampilkan dropdown buku hanya yang ada stock lebih dari 1

michaelr1300 commented on application/modules/invoice/models/invoice_model.php:
```diff
@@ -0,0 +1,326 @@
+        return $this->db->get()->row();
+    }
+
+    public function get_last_invoice_number($type)
```
Function untuk generate nomor faktur

bagaskarala commented:
ada error ketika klik tambah barang

![image](https://user-images.githubusercontent.com/14112952/117105347-08485080-ada8-11eb-90e2-6d505b62259f.png)

bagaskarala commented:
bagian ini agak ambigu, pilih customer tapi bisa juga create customer.
saranku mungkin dibikin tab aja, pilih customer atau masukkan customer baru

![image](https://user-images.githubusercontent.com/14112952/117105502-5cebcb80-ada8-11eb-833a-e9ca9aca7435.png)

bagaskarala commented:
view `add_invoice` dan `edit_invoice` kalau alurnya sama lebih baik dijadikan satu aja `form_invoice` biar tidak redundan. nanti perbedaannya diatur berdasarkan url `invoice/add` atau `invoice/edit`.

bagaskarala commented:
error jika milih non-tunai, sehingga tidak mengisi asal stok.
column `source` di db perlu nullable ?

![image](https://user-images.githubusercontent.com/14112952/117106204-a983d680-ada9-11eb-8825-cd93cc280aa5.png)

bagaskarala commented on application/modules/invoice/controllers/invoice.php:
```diff
@@ -0,0 +1,293 @@
+    }
+
+    // Auto generate nomor faktur berdasar jenis faktur
+    public function api_get_last_invoice_number($type)
+    {
+        $number = $this->invoice->get_last_invoice_number($type);
+        return $this->send_json_output(true, $number);
+    }
+}
```
ini kenapa perlu dibikin API?
gak diproses ketika submit aja?

michaelr1300 replied:
Oke mas, sudah diganti, sekarang diproses di model saat submit

michaelr1300 committed:
Changes & Bug Fix

- Migration: Edit kolom source menjadi Nullable
- UI changes: Pemilihan Customer dipisah dengan Tab
- Controller: Remove api_get_last_invoice_number
- Model: Get last invoice number diproses saat submit

michaelr1300 commented:
> view `add_invoice` dan `edit_invoice` kalau alurnya sama lebih baik dijadikan satu aja `form_invoice` biar tidak redundan. nanti perbedaannya diatur berdasarkan url `invoice/add` atau `invoice/edit`.

Yang ini rencananya akan ada perbedaan sedikit mas untuk view edit dan add. Nantinya edit tidak bisa ubah semua data faktur, jenis faktur tidak bisa di edit. Sekarang ini sementara masih yang semuanya mas

michaelr1300 commented:
> ada error ketika klik tambah barang
> 
> ![image](https://user-images.githubusercontent.com/14112952/117105347-08485080-ada8-11eb-90e2-6d505b62259f.png)

Oke mas, noted. Sudah di fix Jquery nya jadi tidak kedobel manggilnya

bagaskarala commented:
> > view `add_invoice` dan `edit_invoice` kalau alurnya sama lebih baik dijadikan satu aja `form_invoice` biar tidak redundan. nanti perbedaannya diatur berdasarkan url `invoice/add` atau `invoice/edit`.
> 
> Yang ini rencananya akan ada perbedaan sedikit mas untuk view edit dan add. Nantinya edit tidak bisa ubah semua data faktur, jenis faktur tidak bisa di edit. Sekarang ini sementara masih yang semuanya mas

oke sipp
